README.md
At difficulty 400000 ~3.5 TPS

Goal: create test with 3 test cases


node1/history
miner.start
miner.start()
miner.stop()
miner.start()
miner.stop()
miner.start()
miner.stop()
miner.start()
miner.stop()
web3.eth.getTransactionCount(fromAddress, 'pending')
web3.eth.getTransactionCount('pending')
web.eth.getTransactionCount()
web3.eth.getTransactionCount()
miner.start()

test.js
const Web3 = require('web3');
const web3 = new Web3(new Web3.providers.HttpProvider('http://localhost:8000'));

const keythereum = require('keythereum');
const keyobj = keythereum.importFromFile('0x05c7A90D29D2e43B53dad43E64d71405BaECaCA9', './node1');
const PRIVATE_KEY = keythereum.recover('1234', keyobj).toString('hex');

function timeout(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function testChain(numberOfTransactions, TPS) {
  let n = 0;
  while (n < numberOfTransactions) {
    const curNonce = await web3.eth.getTransactionCount('0x05c7A90D29D2e43B53dad43E64d71405BaECaCA9', 'pending');
    await Promise.all(
      Array(25).fill().map(async (_, i) =>
        web3.eth.accounts.signTransaction({
          data: web3.utils.toHex(`dasdasdsdfd${n}`),
          gas: '1000000',
          nonce: curNonce + i,
        }, PRIVATE_KEY).then(({ rawTransaction }) => 
          web3.eth.sendSignedTransaction(rawTransaction, function(error, hash) {
            if (!error) {
              console.log("🎉 The hash of your transaction is: ", hash);
            } else {
              console.log("❗Something went wrong while submitting your transaction:", error)
            }
            }
          )
        )
      )
    )
  }
}

testChain();

index.js
const java = require('java');
const Web3 = require('web3');
const web3 = new Web3(new Web3.providers.HttpProvider('http://localhost:8000'));
const path = require('path');

const jarPath = path.join(__dirname, 'ChameleonHash.jar');
java.classpath.push(jarPath);

const KEY_SIZE = 2048;

const node1 = '';
const wallet1 = '';

const node2 = '';
const wallet2 = '';

const ChameleonHash = java.import('ChameleonHash');
const BigInteger = java.import('java.math.BigInteger');

const m = new BigInteger('123456789');
const pp = java.newArray('java.math.BigInteger', ChameleonHash.setupSync(KEY_SIZE)); // Cast into BigInteger[]

const key = ChameleonHash.keyGenSync(pp[0], pp[1], pp[2], KEY_SIZE);
const sk = key[0];
const pk = key[1];
console.log(pp);
const hash = ChameleonHash.hashSync(m, pk, pp);
console.log(hash[0].toStringSync());



start.sh
geth --genesis genesis.json --datadir node --networkid 1900 --nodiscover --maxpeers 0 console

geth --datadir node1 account new
geth --datadir node1 init genesis.json
geth --datadir node2 init genesis.json
geth --identity node1 --http --http.port "8000" --http.corsdomain "*" --datadir node1 --port "30303" --nodiscover --http.api "db,eth,net,web3,personal,miner,admin" --networkid 1900 --nat "any" --allow-insecure-unlock console
geth --identity node2 --http --http.port "8001" --http.corsdomain "*" --datadir node2 --port "30304" --nodiscover --http.api "db,eth,net,web3,personal,miner,admin" --networkid 1900 --nat "any" --allow-insecure-unlock console

admin.addPeer("enode://b20fe9866aa54bc3c44f3bbb575f3f253b6cf395b2285c71528e9dc422765a51f3c863e117ddc817f2922e1535a46c7e2c05e45527ea3bc16ad06a2e14d2b580@[::]:30304?discport=0")
admin.addPeer("enode://f8adf8b239a821e0d434c6eb9c79f636ba5a995d1f3b30bb964daa868d5ead9e38537cd9610f6833a39bd905660410a3b2fa90297bbdb49235a680fc0a8dd97b@[::]:30303?discport=0")

testCH.java
import java.math.BigInteger;

public class testCH {
	public static void main(String[] args) {
		try {
		BigInteger[] pp = ChameleonHash.setup(2048);
		BigInteger[] key = ChameleonHash.keyGen(pp[0], pp[1], pp[2], 2048);
 		
 		BigInteger m = new BigInteger("1234567890123456789");
 		BigInteger sk = key[0];
 		BigInteger pk = key[1];
 		BigInteger[] ch = ChameleonHash.hash(m, pk, pp);

 		System.out.println("m = " + ch[0].toString());
 		System.out.println("r = " + ch[1].toString());
 		System.out.println("v = " + ch[2].toString());
 		if (ChameleonHash.verify(pk, ch, pp))
 			System.out.println("Chameloen hash is valid.");
 		else
 			System.out.println("Chameloen hash is invalid.");

 		BigInteger mPrime = new BigInteger("9876543210987654321");
 		ChameleonHash.adapt(mPrime, sk, ch, pp);

 		System.out.println("mPrime = " + ch[0].toString());
 		System.out.println("rPrime = " + ch[1].toString());
 		System.out.println("v = " + ch[2].toString());
 		if (ChameleonHash.verify(pk, ch, pp))
 			System.out.println("Chameloen hash is valid.");
 		else
 			System.out.println("Chameloen hash is invalid.");
		}
		catch (Exception e) {
			System.out.println(e);
		}
	}
}

ChameleonHash.java
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.math.BigInteger;
import java.security.interfaces.DSAPublicKey;

public class ChameleonHash {
	public static BigInteger[] setup(Integer bitLength) throws Exception {
		KeyPairGenerator keyGen = KeyPairGenerator.getInstance("DSA", "SUN");
        SecureRandom random = SecureRandom.getInstance("SHA1PRNG", "SUN");
        keyGen.initialize(bitLength, random);
        KeyPair pair = keyGen.generateKeyPair();
        PublicKey pub = pair.getPublic();
        DSAPublicKey dsaPublicKey = (DSAPublicKey) pub;

        BigInteger[] pp = new BigInteger[4];	// g, p, q, bitLength
        pp[0] = dsaPublicKey.getParams().getG();
        pp[1] = dsaPublicKey.getParams().getP();
        pp[2] = dsaPublicKey.getParams().getQ();
        pp[3] = new BigInteger(Integer.toString(bitLength));

        return pp;
	}

	public static BigInteger[] keyGen(BigInteger g, BigInteger p, BigInteger q, int bitLength) throws Exception {
		BigInteger[] key = new BigInteger[2];	// sk, pk
		SecureRandom sr = SecureRandom.getInstance("SHA1PRNG", "SUN");
        BigInteger sk = new BigInteger(bitLength, sr).mod(q);
        BigInteger pk = g.modPow(sk, p);

        key[0] = sk;
        key[1] = pk;

        return key;
	}

	public static BigInteger[] hash(BigInteger m, BigInteger pk, BigInteger[] pp) throws Exception {
		BigInteger[] chValues = new BigInteger[3];	// message, rand and hashValue
		BigInteger g = pp[0];
		BigInteger p = pp[1];
		int bitLength = pp[3].intValue();
		BigInteger r = new BigInteger(bitLength, SecureRandom.getInstance("SHA1PRNG", "SUN"));
		BigInteger v = g.modPow(m,p).multiply(pk.modPow(r,p)).mod(p);
		chValues[0] = m;
		chValues[1] = r;
		chValues[2] = v;

		return chValues;
	}

	public static boolean verify(BigInteger pk, BigInteger[] chValues, BigInteger[] pp) {
		BigInteger m = chValues[0];
		BigInteger r = chValues[1];
		BigInteger v = chValues[2];
		BigInteger g = pp[0];
		BigInteger p = pp[1];

		BigInteger vPrime = g.modPow(m,p).multiply(pk.modPow(r,p)).mod(p);

		if (vPrime.equals(v))
			return true;
		return false;
	}

	public static void adapt(BigInteger mPrime, BigInteger sk, BigInteger[] chValues, BigInteger[] pp) {
		BigInteger g = pp[0];
		BigInteger p = pp[1];
		BigInteger q = pp[2];
		BigInteger m = chValues[0];
		BigInteger r = chValues[1];
		BigInteger rPrime = m.add(sk.multiply(r)).add(mPrime.negate()).multiply(sk.modInverse(q)).mod(q);
		chValues[0] = mPrime;
		chValues[1] = rPrime;
	}
}

blockchainInterface.js
const Web3 = require('web3');
const { hrtime } = require('process');
const NS_PER_SEC = 1e9;

const web3 = new Web3(new Web3.providers.HttpProvider('http://localhost:8000'));

function timeout(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

let address;
let privateKey;

function init(_address, _privateKey) {
  address = _address;
  privateKey = _privateKey;
}

// Send 50 transactions and see how quickly they get confirmed
async function calculate50TPS() {
  const pendingTransactionCount = await web3.eth.getTransactionCount(address, 'pending');
  const startTransactionCount = await web3.eth.getTransactionCount(address);

  const NUM_OF_TRANSACTIONS = 50;

  const start = hrtime();

  for (let n = 0; n < NUM_OF_TRANSACTIONS; n++) {
    web3.eth.accounts.signTransaction({
      data: web3.utils.toHex('Hello world'),
      gas: '1000000',
      nonce: pendingTransactionCount + n,
    }, privateKey).then(({ rawTransaction}) => web3.eth.sendSignedTransaction(rawTransaction, function(error, hash) {
      if (!error) {
        console.log('🎉 Transaction created');
      } else {
        console.log(error)
      }
    }));
  }

  let currentTransactionCount = await web3.eth.getTransactionCount(address);
  while (currentTransactionCount < startTransactionCount + NUM_OF_TRANSACTIONS) { 
    console.log('Retrying... ' + currentTransactionCount)
    await timeout(200)
    currentTransactionCount = await web3.eth.getTransactionCount(address); 
  }

  const diff = hrtime(start);
  const timeTaken = diff[0] + (diff[1] / NS_PER_SEC); 

  return NUM_OF_TRANSACTIONS / timeTaken;
}

async function testChain(numberOfTransactions, TPS) {
  console.log(`Starting test with ${numberOfTransactions} transactions at roughly ${TPS} TPS`);
  const timeBetweenTransactions = 1000 / TPS;
  let droppedTransactions = 0;

  const startTransactionCount = await web3.eth.getTransactionCount(address); 
  const startPendingTransactionCount = await web3.eth.getTransactionCount(address, 'pending');
  
  const start = hrtime();
  for (let n = 0; n < numberOfTransactions; n++) {
    web3.eth.accounts.signTransaction({
      data: web3.utils.toHex('Hello world'),
      gas: '1000000',
      nonce: startPendingTransactionCount + n,
    }, privateKey).then(({ rawTransaction}) => web3.eth.sendSignedTransaction(rawTransaction, function(error, hash) {
      if (!error) {
        // console.log('🎉 Transaction created');
      } else {
        console.log(error)
      }
    }))
    .catch((error) => { console.log('Transaction failed'); droppedTransactions++; })

    await timeout(timeBetweenTransactions);
  }

  const sentDiff = hrtime(start);
  const sentTimeTaken = sentDiff[0] + (sentDiff[1] / NS_PER_SEC);
  console.log(`Total number of dropped transactions: ${droppedTransactions}`);
  console.log(`Sent TPS: ${(numberOfTransactions - droppedTransactions) / sentTimeTaken}`);
  
  console.log('Waiting for all transactions to be confirmed...')
  let currentTransactionCount = await web3.eth.getTransactionCount(address);
  while (currentTransactionCount < startTransactionCount + numberOfTransactions) { 
    await timeout(200);
    currentTransactionCount = await web3.eth.getTransactionCount(address); 
  }

  const confirmedDiff = hrtime(start);
  const confirmedTimeTaken = confirmedDiff[0] + (confirmedDiff[1] / NS_PER_SEC);

  console.log(`Confirmed TPS: ${(numberOfTransactions - droppedTransactions) / confirmedTimeTaken}`);
}

module.exports = {
  init,
  timeout,
  calculate50TPS,
  testChain,
};


blockchain.js
const { spawnSync, spawn } = require('child_process');
const fs = require('fs');
const keythereum = require('keythereum');

const DATADIR = './tmp/node';
const PASSWORD = '1234';
const PASSWORD_FILE = './tmp/password';
const NODE_LOGFILE = './tmp/node.log';
const GENESIS_FILE = './tmp/genesis.json';
const GENESIS = {
  "config": {
    "chainId": 15,
    "homesteadBlock": 0,
    "eip150Block": 0,
    "eip155Block": 0,
    "eip158Block": 0
  },
  "difficulty": "400000",
  "gasLimit": "0x8000000",
  "coinbase": "0x0000000000000000000000000000000000000000",
  "alloc": {},
};

// Create and initialize node with account
// Returns the address and private key
function initAndStartBlockchain() {
  spawnSync('rm', ['-rf', './tmp']);
  spawnSync('mkdir', ['./tmp']);
  spawnSync('mkdir', [DATADIR]);
  spawnSync('touch', [PASSWORD_FILE]);
  fs.writeFileSync(PASSWORD_FILE, PASSWORD);
  spawnSync('geth', ['--datadir', DATADIR, 'account', 'new', '--password', PASSWORD_FILE]);

  const accounts = fs.readdirSync(`${DATADIR}/keystore`);
  const address = `0x${JSON.parse(fs.readFileSync(`${DATADIR}/keystore/${accounts[0]}`, 'utf8')).address}`;

  GENESIS.alloc[address] = { balance: '99999999999999999999999999999999' };

  fs.writeFileSync(GENESIS_FILE, JSON.stringify(GENESIS, null, 4));
  spawnSync('geth', ['--datadir', DATADIR, 'init', GENESIS_FILE]);

  const logStream = fs.createWriteStream(NODE_LOGFILE, { flags: 'a' });

  const node = spawn('geth', [
    '--datadir', DATADIR, '--http', '--http.port', '8000', '--nodiscover', '--http.api',
    '\"db,eth,net,web3,personal,miner,admin\"', '--networkid', '1900', '--port', '33333', '--mine', '--miner.threads=1'
  ]);

  node.stdout.pipe(logStream);
  node.stderr.pipe(logStream);
  node.on('close', function (code) {
    console.log('geth node exited with code ' + code);
  });

  process.on('exit', function() {
    node.kill()
  });

  const keyobj = keythereum.importFromFile(address, DATADIR);
  const privateKey = keythereum.recover(PASSWORD, keyobj).toString('hex');

  return {
    address,
    privateKey,
  };
}

module.exports = initAndStartBlockchain;
